malloc/malloc.c

```diff
@@ -3,47 +3,69 @@
 #include "lib.h"
 
 /*
-    Duplicates the input string by dynamically allocating memory for 
+    Duplicates the input string by dynamically allocating memory for
     the duplicate string using `malloc` and then copying the string
     into the allocated memory. Returns a pointer to the allocated memory.
     You may want to use the string_length function to figure out the
     length of the input string.
-    
+
     Do not use the `strdup` function from the standard library.
 */
 char *string_dup(char *src)
 {
-
+    int n = string_length(src);
+    char *dup = malloc(n + 1);
+    for (int i = 0; i < n; i++)
+    {
+        *(dup + i) = *(src + i);
+    }
+    *(dup + n) = '\0';
+    return dup;
 }
 
 /*
     A generic version of string_copy, mem_copy receives a block of memory
     of any type and copies its contents to the destination pointer (dest).
-    `n` is the number of bytes to copy. You may want to cast the input 
-    pointers to char pointers first before performing the copying. 
-    
+    `n` is the number of bytes to copy. You may want to cast the input
+    pointers to char pointers first before performing the copying.
+
     Do not use the `memcpy` function from the standard library.
 */
 void *mem_copy(void *dest, const void *src, int n)
 {
-
+    char *copy_src = src;
+    char *copy_dest = dest;
+    for (int i = 0; i < n; i++)
+    {
+        *(copy_dest + i) = *(copy_src + i);
+    }
 }
 
 /*
-    Given a pointer that `malloc`'d memory, this function will 
+    Given a pointer that `malloc`'d memory, this function will
     attempt to resize the allocated memory to the new specified
-    size. Any data that was previously in the old `malloc`'d 
-    memory should be intact in the new resized block of memory. 
+    size. Any data that was previously in the old `malloc`'d
+    memory should be intact in the new resized block of memory.
 
     Some edge cases to consider: how should resizing be handled
-    in the case when old_size < new_size? What about when 
+    in the case when old_size < new_size? What about when
     old_size > new_size?
 
     Do not use the `realloc` function from the standard libary.
 */
 void *resize_memory(void *ptr, int old_size, int new_size)
 {
-
+    void *new_block = malloc(new_size);
+    if (new_size < old_size)
+    {
+        mem_copy(new_block, ptr, new_size);
+    }
+    else
+    {
+        mem_copy(new_block, ptr, old_size);
+    }
+    free(ptr);
+    return new_block;
```

Comment:
The test for resize memory isn't passing. The tests are checking that if the existing memory has a string and you resize it to be less than it's original size, the string in the resized memory is properly truncated (i.e. is properly terminated with a `'\0'`.